pennylane/beta/tapes/tape.py

```diff
@@ -936,3 +936,387 @@ def execute_device(self, params, device):
+               [ 0.00079144,  0.00103601,  0.00737611],
+               [ 0.178441  ,  0.23358253, -0.00737611]])
+
+        Parameters can be optionally passed during execution:
+
+        >>> tape.jacobian(dev, params=[1.0, 0.0, 1.0])
+        array([[-3.24029934e-01, -9.99200722e-09, -3.24029934e-01],
+               [-9.67055711e-02, -2.77555756e-09,  3.24029935e-01],
+               [ 9.67055709e-02,  3.05311332e-09,  9.67055709e-02],
+               [ 3.24029935e-01,  1.08246745e-08, -9.67055711e-02]])
+
+        Parameters provided for execution are temporary, and do not affect
+        the tapes' parameters in-place:
+
+        >>> tape.get_parameters()
+        [0.432, 0.543, 0.133]
+
+        Explicitly setting the trainable parameters can significantly reduce
+        computational resources, as non-trainable parameters are ignored
+        during the computation:
+
+        >>> tape.trainable_params = {0} # set only the first parameter as trainable
+        >>> tape.jacobian(dev)
+        array([[-0.178441  ],
+               [-0.00079144],
+               [ 0.00079144],
+               [ 0.178441  ]])
+
+        If a tape has no trainable parameters, the Jacobian will be empty:
+
+        >>> tape.trainable_params = {}
+        >>> tape.jacobian(dev)
+        array([], shape=(4, 0), dtype=float64)
+        """
+        method = options.get("method", "best")
+
+        if method not in ("best", "numeric", "analytic", "device"):
+            raise ValueError(f"Unknown gradient method '{method}'")
+
+        if params is None:
+            params = self.get_parameters()
+
+        params = np.array(params)
+
+        if method == "device":
+            # Using device mode; simply query the device for the Jacobian
+            return self.device_pd(device, params=params, **options)
+
+        # perform gradient method validation
+        allowed_param_methods = self._grad_method_validation(method)
+
+        if not params.size or all(g == "0" for g in allowed_param_methods):
+            # Either all parameters had grad method 0, or there are no trainable
+            # parameters. Simply return an empty Jacobian.
+            return np.zeros((self.output_dim, len(params)), dtype=float)
+
+        if method == "numeric" or "F" in allowed_param_methods:
+            # there exist parameters that will be differentiated numerically
+
+            if options.get("order", 1) == 1:
+                # First order (forward) finite-difference will be performed.
+                # Compute the value of the tape at the current parameters here. This ensures
+                # this computation is only performed once, for all parameters.
+                options["y0"] = np.asarray(self.execute_device(params, device))
+
+        jac = None
+        p_ind = list(np.ndindex(*params.shape))
+
+        # loop through each parameter and compute the gradient
+        for idx, (l, param_method) in enumerate(zip(p_ind, allowed_param_methods)):
```

Comment:
@mariaschuld: this is the for loop we want to avoid performing in serial for devices that support batch execution. Instead, we would want to:

* 'accumulate' all the tapes generated by this for loop, and submit that at once via `device.batch_execute()`
* post-process the result to create the jacobian matrix.

Alternatively, we could simply 'dask' this for loop, but I am a bit uneasy re: race conditions.

Reply:
Actually, just had a thought. We could simply do `device.batch_execute(tape, list_of_params)` (pass a single tape with a set of batched parameters, rather than multiple tapes).

While less 'elegant', it might have less overhead.

Reply:
Good point. Could we misuse this PR to prepare batching, once we are at it? And also to support gates with more than 2 shifts that may have distinct weighing terms for every shift?

My current hope is that a single tape could represent multiple circuits, but it is not yet clear to me how this will look like.

But to be very flexible, I would propose to rewrite the diff methods so that they collect a list of pairs ``(shift, weight)`` for every trainable parameter/index. The nice thing is that both analytic and finite differences are of that form, and it is a compact representation of the $\sum_i |shifts_i|$ circuits that result from shifting any single parameter. 

In future there might be optimisers that need access to this list to find shortcuts to running all those circuits (i.e., imagine that all shifts must be applied at once...kind of like SPSA @antalszava ?). So it would be good to design it so that a function can be applied to this "shift bookkeeping datastructure".

Then, somewhere central, the shift bookkeeping data structure can be turned into actual device evaluation - we'd have the freedom to change that part once it's clear how batch evaluation works....

Does that make sense?